goss-testing/automated/run-ncn-tests.sh

```diff
@@ -45,6 +45,22 @@ function is_pit_node {
     return $?
 }
 
+# Set some default variables here, both because we use them in this library, and also to save
+# scripts from having to set them if they source this library.
+
+GOSS_INSTALL_BASE_DIR="/opt/cray/tests/install"
+
+if [[ -z ${GOSS_BASE} ]]; then
+    if is_pit_node ; then
+        GOSS_BASE="${GOSS_INSTALL_BASE_DIR}/livecd"
+    else
+        GOSS_BASE="${GOSS_INSTALL_BASE_DIR}/ncn"
+    fi
+fi
+export GOSS_BASE
```

Comment:
Shouldn't this be up a line? if the -z doesn't pass this could set GOSS_BASE to null and export it into the env.

Reply:
If the -z doesn't pass, that would mean it was already set. In that case, I believe this export command would export it with its current value. Is that not correct?

Reply:
True, though it could be null and that still passes it through though for that it might be better to do export GOSS_BASE=${GOSS_BASE:?} so if its set and null or unset it exits so that things can't trickle down further. All depends on if a set and null value should(n't) get passed into the env of anything fork()/exec()'d later.

Reply:
Do you mean null as it just whitespace? I think if it's null as in no value at all, -z would fail, wouldn't it?

Reply:
```
root@jolt1-ncn-m001-pit 2022-08-01 20:34:54 /var/log/conman # D=""
root@jolt1-ncn-m001-pit 2022-08-01 20:39:29 /var/log/conman # [[ -z ${D} ]]
root@jolt1-ncn-m001-pit 2022-08-01 20:39:36 /var/log/conman # echo $?
0
```

Reply:
Not exactly (bash but all posix sh should behave the same):

```$ (export FOO=; [ -z $FOO ] && echo set and null || echo not ok )
set and null
$ (export FOO=blah; [ -z $FOO ] && echo set and not null || echo not ok )
not ok``

Reply:
I'm confused here. When you have it set and null, the -z is returning return code 0, right? Which means that the if stanza in question here would have set the variable, wouldn't it?

Reply:
Sorry forgot the third possibility:

```
$ ([ -z $FOO ] && echo not set || echo not ok)
not set
```

Reply:
This is the code in question:

```
if [[ -z ${GOSS_BASE} ]]; then
    if is_pit_node ; then
        GOSS_BASE="${GOSS_INSTALL_BASE_DIR}/livecd"
    else
        GOSS_BASE="${GOSS_INSTALL_BASE_DIR}/ncn"
    fi
fi
export GOSS_BASE
```

If GOSS_BASE starts out set and null, or not set at all, either way it will get set inside the if clause, won't it?

Reply:
Although thinking about it you're right, ignore this chain, my brains becoming a pumpkin.

Reply:
I'm trying to figure out what starting situation would result in GOSS_BASE being exported with a null value

Reply:
Whew. Okay. I thought I was losing my mind. :) I mean, moreso than usual.